File: src/main/java/com/hileco/cortex/vm/VirtualMachine.java
package com.hileco.cortex.vm;

import com.hileco.cortex.vm.layer.LayeredMap;
import com.hileco.cortex.vm.layer.LayeredStack;
import lombok.Data;

import java.math.BigInteger;

@Data
public class VirtualMachine {
    public static final int INSTRUCTION_LIMIT = 1_000_000;
    public static final BigInteger NUMERICAL_LIMIT = new BigInteger(new byte[]{2}).pow(256).subtract(BigInteger.ONE);

    private LayeredStack<ProgramContext> programs;
    private LayeredMap<BigInteger, Program> atlas;
    private BigInteger overflowLimit;
    private BigInteger underflowLimit;
    private long stackLimit;
    private int instructionsExecuted;
    private int instructionLimit;

    public VirtualMachine(ProgramContext... programContexts) {
        this.programs = new LayeredStack<>();
        for (var programContext : programContexts) {
            this.programs.push(programContext);
        }
        this.atlas = new LayeredMap<>();
        this.stackLimit = Long.MAX_VALUE;
        this.overflowLimit = NUMERICAL_LIMIT;
        this.underflowLimit = NUMERICAL_LIMIT;
        this.instructionsExecuted = 0;
        this.instructionLimit = INSTRUCTION_LIMIT;
    }
}

File: src/main/java/com/hileco/cortex/vm/VirtualMachine.kt
package com.hileco.cortex.vm

import com.hileco.cortex.vm.layer.LayeredMap
import com.hileco.cortex.vm.layer.LayeredStack
import java.math.BigInteger

class VirtualMachine(vararg programContexts: ProgramContext) {
    val programs: LayeredStack<ProgramContext> = LayeredStack()
    val atlas: LayeredMap<BigInteger, Program> = LayeredMap()
    val overflowLimit: BigInteger = NUMERICAL_LIMIT
    val underflowLimit: BigInteger = NUMERICAL_LIMIT
    val stackLimit: Long = java.lang.Long.MAX_VALUE
    var instructionsExecuted: Int = 0
    val instructionLimit: Int = INSTRUCTION_LIMIT

    init {
        for (programContext in programContexts) {
            programs.push(programContext)
        }
    }

    fun dump() {
        var programIndex = 0
        for (program in programs) {
            var stackIndex = 0
            val iterator = program.stack.iterator()
            while (iterator.hasNext()) {
                val element = iterator.next()
                val bigInteger = BigInteger(element)
                println("{$programIndex} [$stackIndex] = $bigInteger")
                stackIndex++
            }
            programIndex++
        }
    }

    companion object {
        const val INSTRUCTION_LIMIT = 1000000
        val NUMERICAL_LIMIT: BigInteger = BigInteger(byteArrayOf(2)).pow(256).subtract(BigInteger.ONE)
    }
}
